src/coloring.rs

```diff
@@ -63,22 +79,29 @@ use rustworkx_core::coloring::{
 /// .. [1] Adrian Kosowski, and Krzysztof Manuszewski, Classical Coloring of Graphs,
 ///     Graph Colorings, 2-19, 2004. ISBN 0-8218-3458-4.
 #[pyfunction]
-#[pyo3(text_signature = "(graph, /, preset_color_fn=None)")]
+#[pyo3(text_signature = "(graph, /, preset_color_fn=None, greedy_strategy=GreedyStrategy::Degree)")]
+#[pyo3(signature=(graph, /, preset_color_fn=None, greedy_strategy=GreedyStrategy::Degree))]
 pub fn graph_greedy_color(
     py: Python,
     graph: &graph::PyGraph,
     preset_color_fn: Option<PyObject>,
+    greedy_strategy: GreedyStrategy,
 ) -> PyResult<PyObject> {
+    let inner_strategy = match greedy_strategy {
+        GreedyStrategy::Saturation => GreedyStrategyCore::Saturation,
+        GreedyStrategy::Degree => GreedyStrategyCore::Degree,
+    };
+
     let colors = match preset_color_fn {
         Some(preset_color_fn) => {
             let callback = |node_idx: NodeIndex| -> PyResult<Option<usize>> {
                 preset_color_fn
                     .call1(py, (node_idx.index(),))
                     .map(|x| x.extract(py).ok())
             };
-            greedy_node_color_with_preset_colors(&graph.graph, callback)?
+            greedy_node_color_with_preset_colors(&graph.graph, callback, inner_strategy)?
         }
-        None => greedy_node_color(&graph.graph),
+        None => greedy_node_color(&graph.graph, inner_strategy),
     };
```

Comment:
I was wondering if we really need two different functions `greedy_node_color` and `greedy_node_color_with_preset_colors` which are both simple wrappers on top of `inner_greedy_node_color`. I.e. maybe we could have a single function, with the preset function specified via an `Option<...>`. 

One possible complication is that when the preset function is specified, the coloring algorithm may technically fail as we may specify the same color to two nodes connected by an edge. Though, I don't think we actually check for this, should we? Or maybe this behavior should be allowed, i.e. the user may specify an inconsistent preset fn if the goal is to only greedily color the remaining nodes.

Reply:
This was for backwards compatibility between subsequent rustworkx-core releases. There isn't an easy way to add optional parameters to a rust function, so when I added the `preset_colors` option in #1060 I opted to make it a new function so that any existing users of `graph_node_color` were not broken by needed to pass a new `None` argument to `graph_node_color` when they upgraded to `rustworkx-core` 0.14.0. On the python side I did expose it using the same entry point though, but for the pure rust library compatibility I didn't have a good pattern for doing this. (there might be one that I'm ignorant of though).

As for the soundness, yeah it is likely that the preset color function could result in an incorrect or invalid coloring. The idea was to do as you described to enable greedily coloring the remaining nodes, or at least the was the use case I had in mind (and was asked for an implementation of). But we don't check that the end result coloring between the user provided function provides a globally valid coloring. I tried to document this as a potential issue in the documentation where I wrote:

> Note if you do use a callable there is no validation that the preset values are valid colors. You can generate an invalid coloring if you the specified function returned invalid colors for any nodes.

but if you think we should make the language there stronger I think it's a good idea. TBH, this feature felt like a fairly niche use case I didn't expect it to come up in practice very much.